### PR TITLE
Three-state labels

### PR DESCRIPTION
**Pull request by nknguyenhc**
### Summary:

Fixes part of #240.

#### Type of change:

- ✨ New Feature/ Enhancement

### Changes Made:

Made the filter labels a three-state button.

Each filter label has the following 3 possible states:
* Default: no filter applied. This corresponds to no border.
* Selected: filter **in** all issues/PRs with this label. This corresponds to green border.
* Deselected: filter **out** all issues/PRs with this label. This corresponds to red border.

Additionally, each label has the following state rotation: default -> selected -> deselected.

The old checkbox for two-state is removed. Labels are manually added to the set of selected/deselected label names.

Finally, the set of deselected label names are used in the filter to apply them to the issues/PRs displayed.

There is one case that I need to highlight. That is, if an issue/PR has 2 labels, one is in selected state, and one is in deselected state. Currently, the filter with deselected state takes precedence, so an issue/PR with deselected labels will be filtered out, regardless of whether any other of its labels are being selected. Consider #251 with labels `category.Feature` and `category.Enhancement`, currently being assigned to me:

image

It is being filtered out.

### Screenshots:

image

image

### Proposed Commit Message:

```
Three-state labels

Previously, each label only has 2 states, either selected or
not selected. However, with such design,
the feature of hiding labels can be confused
with hiding issues/PRs with the label.

We implement the three-state label filters,
so that each label can also be used to hide
issues/PRs with the label.
```

<details><summary>
<h3>Checklist:</h3>
</summary>

- [X] I have tested my changes thoroughly.
- [ ] I have created tests for any new code files created in this PR or provided a link to a issue/PR that addresses this.
- [X] I have added or modified code comments to improve code readability where necessary.
- [ ] I have updated the project's documentation as necessary.

</details>

### CODE REVIEW

File: src/app/shared/filter-bar/label-filter-bar/label-filter-bar.component.css [hunk 1]
```diff
@@ -88,17 +88,19 @@
   flex-direction: row;
   justify-content: flex-start;
   align-items: center;
+  width: calc(100% - 8px);
+  border-radius: 10px;
+  height: 40px;
+  padding: 0px 12px;
+  margin: 8px 4px;
+  box-sizing: border-box;
```
Comment (Arif-Khalid) [inline src/app/shared/filter-bar/label-filter-bar/label-filter-bar.component.css:96]: Are the height, width and box-sizing properties necessary? 
The component seems to behave the same without these properties.

**Review by Arif-Khalid — COMMENTED** [review]
Great working implementation of three-state labels! 👏
Just minor comments about code quality but overall a great job!

**Comment by Arif-Khalid** [comment]
As an add-on unrelated to the code but more so the design. 
I feel that the harsh border colors clash with the design of watcher.
To me, a more fitting way of showing the states would be changing the background color of the entire label to a light red/pink for deselected and light green for selected.
image

**Comment by damithc** [comment]
Other possibilities: using ~~strike through~~, an icon/symbol
If using colors to convey information, need to consider color-blind people (i.e., use contrasting colors)

File: src/app/shared/filter-bar/label-filter-bar/label-filter-bar.component.html [hunk 2]
```diff
@@ -19,32 +19,27 @@
...
-            [value]="label.name"
-            [selected]="selectedLabelNames.includes(label.name)"
-            class="list-option"
-            [class.hidden]="filter(input.value, label.name)"
+        <div
+          *ngFor="let label of this.allLabels"
+          class="flexbox-container"
+          (click)="changeLabelState(label)"
+          [class.hidden]="filter(input.value, label.name)"
+          [style]="{ border: '2px solid ' + getBorderColor(label) }"
```
[the remark this answers is not in the document]
Comment (nknguyenhc) [reply src/app/shared/filter-bar/label-filter-bar/label-filter-bar.component.html:27]: I believe there is no difference in which one is a better code design, but I would prefer not putting style-related code into ts files.

**Comment by nknguyenhc** [comment]
With regards to the indication of label states, I have thought about the solution of making the background color. However, remember that Github labels can be of any color, so red tags and green tags might blend in with the background. I would still prefer sticking with the border color, but incorporate strike through for deselected state. Please see updated screenshot in the PR description.

With regards to the color, I have picked from [Material Design color system](https://m2.material.io/design/color/the-color-system.html).

The bug on `Remove all` button has also been fixed.

File: src/app/shared/filter-bar/label-filter-bar/label-filter-bar.component.ts [hunk 3]
```diff
@@ -61,18 +62,36 @@ export class LabelFilterBarComponent implements OnInit, AfterViewInit, OnDestroy
...
+   */
+  getBorderColor(label: SimpleLabel): string {
+    if (this.selectedLabelNames.has(label.name)) {
+      return '#2bfc2e';
+    } else if (this.deselectedLabelNames.has(label.name)) {
+      return '#ff0303';
+    } else {
+      return 'transparent';
+    }
+  }
```
[the remark this answers is not in the document]
Comment (nknguyenhc) [reply src/app/shared/filter-bar/label-filter-bar/label-filter-bar.component.ts:93]: Done.

File: src/app/shared/filter-bar/label-filter-bar/label-filter-bar.component.html (hunk 2 above)
[the remark this answers is not in the document]
Comment (Arif-Khalid) [reply src/app/shared/filter-bar/label-filter-bar/label-filter-bar.component.html]: I see where you are coming from and I would do the same as you. 
My concern is following the coding style already established in this project as shown from ``labelService.setLabelStyle``.
We should wait for other's opinion @CATcher-org/active-devs.

File: src/app/shared/filter-bar/label-filter-bar/label-filter-bar.component.css (hunk 1 above)
Comment (Arif-Khalid) [reply src/app/shared/filter-bar/label-filter-bar/label-filter-bar.component.css]: I see your reasoning. 
Makes sense to me.

**Review by Arif-Khalid — APPROVED** [review]
I like the changes made and everything is in working order. LGTM!

**Review by Arif-Khalid — APPROVED** [review]
LGTM!

**Review by Arif-Khalid — APPROVED** [review]
LGTM!

**Pull request merged**